File: python/scenario/_judge/span_utils.py
"""
Shared span utility functions for OpenTelemetry span processing.

Standalone public functions for calculating durations, formatting timestamps,
extracting status indicators, token usage, and cleaning attributes.
Used by both the digest formatter and trace discovery tools.
"""

from datetime import datetime, timezone
from typing import Any, Dict

from langwatch.attributes import AttributeKey
from opentelemetry.sdk.trace import ReadableSpan
from opentelemetry.trace import StatusCode


def hr_time_to_ms(hr_time: int) -> float:
    """
    Converts OpenTelemetry nanosecond timestamp to milliseconds.

    Args:
        hr_time: Time in nanoseconds.

    Returns:
        Time in milliseconds.
    """
    return hr_time / 1_000_000


def calculate_span_duration(span: ReadableSpan) -> float:
    """
    Calculates span duration in milliseconds.

    Args:
        span: The OpenTelemetry span.

    Returns:
        Duration in milliseconds.
    """
    start = span.start_time or 0
    end = span.end_time or 0
    return hr_time_to_ms(end) - hr_time_to_ms(start)


def format_duration(ms: float) -> str:
    """
    Formats a duration in milliseconds to a human-readable string.

    Uses milliseconds for durations under 1 second, seconds with
    2 decimal places otherwise.

    Args:
        ms: Duration in milliseconds.

    Returns:
        Formatted duration string (e.g. "150ms" or "2.50s").
    """
    if ms < 1000:
        return f"{round(ms)}ms"
    return f"{ms / 1000:.2f}s"


def format_timestamp(start_time_ns: int) -> str:
    """
    Formats a nanosecond timestamp as an ISO 8601 string.

    Args:
        start_time_ns: Timestamp in nanoseconds.

    Returns:
        ISO 8601 formatted timestamp string with Z suffix.
    """
    ms = hr_time_to_ms(start_time_ns)
    dt = datetime.fromtimestamp(ms / 1000, tz=timezone.utc)
    return dt.isoformat().replace("+00:00", "Z")


def get_status_indicator(span: ReadableSpan) -> str:
    """
    Returns an error status indicator string for a span.

    Args:
        span: The OpenTelemetry span.

    Returns:
        Error indicator string, or empty string for non-error spans.
    """
    if span.status.status_code == StatusCode.ERROR:
        message = span.status.description or "unknown"
        return f" ⚠️ ERROR: {message}"
    return ""


def get_token_usage(span: ReadableSpan) -> str:
    """
    Returns a token usage string for LLM spans using GenAI OTel semantic
    convention attributes (gen_ai.usage.input_tokens + gen_ai.usage.output_tokens).

    Args:
        span: The OpenTelemetry span.

    Returns:
        Token usage string (e.g. ", 1500 tokens"), or empty string if
        no usage attributes are present.
    """
    attrs = dict(span.attributes) if span.attributes else {}
    input_tokens = attrs.get("gen_ai.usage.input_tokens")
    output_tokens = attrs.get("gen_ai.usage.output_tokens")
    if input_tokens is None and output_tokens is None:
        return ""
    total = (int(str(input_tokens)) if input_tokens is not None else 0) + (
        int(str(output_tokens)) if output_tokens is not None else 0
    )
    return f", {total} tokens"


def clean_attributes(attrs: Dict[str, Any]) -> Dict[str, Any]:
    """
    Removes internal/infrastructure attributes from a span's attribute map.

    Strips the ``langwatch.`` prefix from remaining keys and deduplicates.

    Args:
        attrs: Raw attribute dictionary.

    Returns:
        Cleaned attribute dictionary.
    """
    cleaned: Dict[str, Any] = {}
    seen: set = set()

    excluded_keys = [
        AttributeKey.LangWatchThreadId,
        "langwatch.scenario.id",
        "langwatch.scenario.name",
    ]

    for key, value in attrs.items():
        if key in excluded_keys:
            continue
        clean_key = (
            key.replace("langwatch.", "", 1)
            if key.startswith("langwatch.")
            else key
        )
        if clean_key not in seen:
            seen.add(clean_key)
            cleaned[clean_key] = value

    return cleaned


def get_parent_span_id(span: ReadableSpan) -> int | None:
    """
    Extracts the parent span ID from a ReadableSpan.

    Args:
        span: The OpenTelemetry span.

    Returns:
        Parent span ID, or None if the span has no parent.
    """
    if span.parent is not None:
        return span.parent.span_id  # type: ignore[union-attr]
    return None

File: javascript/src/agents/judge/span-utils.ts
import type { ReadableSpan } from "@opentelemetry/sdk-trace-base";
import { attributes } from "langwatch/observability";

import { deepTransform } from "./deep-transform";
import { truncateMediaUrl, truncateMediaPart } from "./truncate-media";

/**
 * Represents a span node in the hierarchy tree.
 * The `shortId` field is the first 8 hex characters of the span ID,
 * used as a stable reference that LLMs can use directly from the skeleton.
 */
export interface SpanNode {
  span: ReadableSpan;
  children: SpanNode[];
  shortId: string;
}

/**
 * Converts OpenTelemetry high-resolution time to milliseconds.
 */
export function hrTimeToMs(hrTime: [number, number]): number {
  return hrTime[0] * 1000 + hrTime[1] / 1_000_000;
}

/**
 * Formats a duration in milliseconds to a human-readable string.
 * Uses milliseconds for durations under 1 second, seconds with 2 decimal places otherwise.
 */
export function formatDuration(ms: number): string {
  if (ms < 1000) return `${Math.round(ms)}ms`;
  return `${(ms / 1000).toFixed(2)}s`;
}

/**
 * Calculates the duration of a span in milliseconds.
 */
export function calculateSpanDuration(span: ReadableSpan): number {
  return hrTimeToMs(span.endTime) - hrTimeToMs(span.startTime);
}

/**
 * Returns an error status indicator string for a span.
 * Returns empty string for non-error spans.
 */
export function getStatusIndicator(span: ReadableSpan): string {
  if (span.status.code === 2) {
    return ` ⚠️ ERROR: ${span.status.message ?? "unknown"}`;
  }
  return "";
}

/**
 * Returns a token usage string for LLM spans using the GenAI OTel semantic convention
 * attributes (gen_ai.usage.input_tokens + gen_ai.usage.output_tokens).
 * Returns empty string if no usage attributes are present.
 */
export function getTokenUsage(span: ReadableSpan): string {
  const input = span.attributes["gen_ai.usage.input_tokens"];
  const output = span.attributes["gen_ai.usage.output_tokens"];
  if (input == null && output == null) return "";
  const total = (Number(input) || 0) + (Number(output) || 0);
  return `, ${total} tokens`;
}

/**
 * Removes internal/infrastructure attributes from a span's attribute map.
 * Strips the `langwatch.` prefix from remaining keys and deduplicates.
 */
export function cleanAttributes(
  attrs: Record<string, unknown>
): Record<string, unknown> {
  const cleaned: Record<string, unknown> = {};
  const seen = new Set<string>();

  const excludedKeys = [
    attributes.ATTR_LANGWATCH_THREAD_ID,
    "langwatch.scenario.id",
    "langwatch.scenario.name",
  ];

  for (const [key, value] of Object.entries(attrs)) {
    if (excludedKeys.includes(key)) {
      continue;
    }
    const cleanKey = key.replace(/^(langwatch)\./, "");
    if (!seen.has(cleanKey)) {
      seen.add(cleanKey);
      cleaned[cleanKey] = value;
    }
  }
  return cleaned;
}

/**
 * Formats a value for display. Processes media URLs and JSON strings recursively,
 * returning a string representation suitable for trace output.
 */
export function formatValue(value: unknown): string {
  const processed = transformValue(value);
  return typeof processed === "string"
    ? processed
    : JSON.stringify(processed);
}

/**
 * Recursively transforms a value, truncating media content and parsing JSON strings.
 */
export function transformValue(value: unknown): unknown {
  return deepTransform(value, (v) => {
    const mediaPart = truncateMediaPart(v);
    if (mediaPart) return mediaPart;
    if (typeof v !== "string") return v;
    const truncated = truncateMediaUrl(v);
    if (truncated !== v) return truncated;
    if (looksLikeJson(v)) {
      try {
        const parsed = transformValue(JSON.parse(v));
        return JSON.stringify(parsed);
      } catch {
        /* not valid JSON */
      }
    }
    return v;
  });
}

/**
 * Checks if a string looks like it could be valid JSON (starts/ends with {} or []).
 */
export function looksLikeJson(str: string): boolean {
  const t = str.trim();
  return (
    (t.startsWith("{") && t.endsWith("}")) ||
    (t.startsWith("[") && t.endsWith("]"))
  );
}

/**
 * Sorts spans by start time and assigns truncated span IDs (first 8 hex chars).
 * Returns a flat array of SpanNodes with shortId set.
 */
export function indexSpans(spans: ReadableSpan[]): SpanNode[] {
  const sorted = [...spans].sort((a, b) => {
    return hrTimeToMs(a.startTime) - hrTimeToMs(b.startTime);
  });

  return sorted.map((span) => ({
    span,
    children: [],
    shortId: span.spanContext().spanId.slice(0, 8),
  }));
}

/**
 * Extracts the parent span ID from a ReadableSpan, handling both OTel SDK v2
 * (parentSpanId: string) and v1 (parentSpanContext: SpanContext) interfaces.
 */
export function getParentSpanId(span: ReadableSpan): string | undefined {
  if (span.parentSpanId) return span.parentSpanId;
  const legacy = (span as unknown as Record<string, unknown>)
    .parentSpanContext as { spanId?: string } | undefined;
  return legacy?.spanId;
}

/**
 * Builds a parent-child hierarchy from indexed span nodes.
 * Returns the root nodes of the hierarchy tree.
 */
export function buildHierarchy(nodes: SpanNode[]): SpanNode[] {
  const bySpanId = new Map<string, SpanNode>();
  for (const node of nodes) {
    bySpanId.set(node.span.spanContext().spanId, node);
  }

  const roots: SpanNode[] = [];
  for (const node of nodes) {
    const parentId = getParentSpanId(node.span);
    const parent = parentId ? bySpanId.get(parentId) : undefined;
    if (parent) {
      parent.children.push(node);
    } else {
      roots.push(node);
    }
  }
  return roots;
}
